### ISSUE TITLE
[Super AI] Implement Gradient Checkpointing

### ISSUE THREAD

**Issue by MoosaTae** [OpenThaiGPT/openthaigpt-pretraining#116]
Description:
Gradient Checkpointing is one the best techniques to reduce model memory usage when training. In Addtiion, there is notion that gradient checkpointing should only be done to attention calculation NOT the entire model https://arxiv.org/abs/2205.05198

Instructions:
- Create Option to do gradient checkpointing for the entire model
- Create Optiion to do gradient checkpointing for only self-attetion

Deliverables:
- Run on Huaweii Cloud
- PR
